source/opt/const_folding_rules.cpp

```diff
@@ -197,13 +197,25 @@ ConstantFoldingRule FoldCompositeWithConstants() {
     analysis::ConstantManager* const_mgr = context->get_constant_mgr();
     analysis::TypeManager* type_mgr = context->get_type_mgr();
     const analysis::Type* new_type = type_mgr->GetType(inst->type_id());
+    Instruction* type_inst =
+        context->get_def_use_mgr()->GetDef(inst->type_id());
 
     std::vector<uint32_t> ids;
-    for (const analysis::Constant* element_const : constants) {
+    for (uint32_t i = 0; i < constants.size(); ++i) {
+      const analysis::Constant* element_const = constants[i];
       if (element_const == nullptr) {
         return nullptr;
       }
-      uint32_t element_id = const_mgr->FindDeclaredConstant(element_const);
+
+      uint32_t component_type_id = 0;
```

Comment:
We likely have a similar issue related to pointers in another function. Fix or file an issue please.

Reply:
I looked at all of the function in the constant manager.  Others member functions that return an instruction or id all take an optional type id as a parameter already.

I missed this one in an earlier change because I did not look closely enough at the implementation.  The const-to-id map would use the address of the constant for the hash, so I figured two constant that "equal" except for different types would both be in the map.

Can you point to any specific ones that I may have missed?